File: C++/2911-Minimum Changes to Make K Semi-palindromes.cpp
// O(n^3*log(n))
class Solution {
public:
    int memo_dp[201][101]; // set 101 because 1 <= k <= s.length / 2
    int memo_v[201][201];
    string s;
    
    int v(int start, int end){
        if (memo_v[start][end] != -1) return memo_v[start][end];
        int length = end-start+1;
        int res = INT_MAX/2;
        // Only O(log(length)) iterations will not hit the continue statement, 
        // so the time complexity for this function is O(length*log(length))
        for (int d=1; d<length; d++){
            if (length%d != 0) continue;
            int currCount = 0;
            int segments = length/d;
            // O(length) time complexity below
            for (int offset=0; offset<d; offset++){
                for (int segmentIdx=0; segmentIdx < segments/2; segmentIdx++){
                    if (s[start+offset+segmentIdx*d] != s[start+offset+(segments-1-segmentIdx)*d]) currCount++;
                }
            }
            res = min(res, currCount);
        }
        memo_v[start][end] = res;
        return res;
    }
    
    int dp(int start, int k){
        if (memo_dp[start][k] != -1) return memo_dp[start][k];
        
        if (k == 1) return v(start, s.size()-1);
        
        int res = INT_MAX/2;
        // Note that there are at least two characters in a semi-palindrome
        for (int mid = start+1; mid <= s.size()-2; mid++){
            res = min(res, v(start, mid) + dp(mid+1, k-1));
        }
        memo_dp[start][k] = res;
        return res;
    }
    
    int minimumChanges(string s, int k) {
        memset(memo_dp, -1, sizeof(memo_dp));
        memset(memo_v, -1, sizeof(memo_v));
        this->s = s;
        return dp(0, k);
    }
};

File: 2911-Minimum Changes to Make K Semi-palindromes.py
# O(n^3*log(n))
class Solution:
    def minimumChanges(self, s: str, k: int) -> int:
        # dp[i][j] = min( v[i][x] + dp[x+1][j] )
        # dp[i][j] is the minimum number of letter changes to change suffix s[i:] into j valid semi-palidromes
        # v[i][x] is the minimum number of letter changes to change substring s[i..x] into semi-palindrome
        # note that there are at least two characters in a semi-palindrome
        
        N = len(s)
        
        @cache
        def dp(start, k):
            if k == 1:
                return v(start, N-1)
            
            res = float('inf')
            for end in range(start+1, N-2):
                res = min(res, v(start, end) + dp(end+1, k-1))
            return res
            
        @cache
        def v(start, end):
            minCount = float('inf')
            length = end-start+1
            
            for d in range(1, length):
                if length%d != 0:
                    continue
                count = 0
                substrings = length//d
                
                for offset in range(d):
                    for i in range(substrings//2):
                        if s[start + offset + i*d] != s[start + offset + (substrings-1-i)*d]:
                            count += 1
                
                minCount = min(minCount, count)
            
            return minCount
        
        return dp(0, k)